Add 2016 Canada census data

antoniocarlon opened the pull request:
Adding Canada 2016 Geographies and Census data.

Closes https://github.com/CartoDB/bigmetadata/issues/324

**Notes for the acceptance:**
 - You can test all the tasks doing `make docker-ca-all-2016`
 - For 2016 the data is not splitted in Census + NHS (currently Census data includes both)
 - I haven't tried each task and some of them will fail with `The following columns of the table "XXX" contain only NULL values` for sure (we will need to add exceptions for them in the `census_columns_filter.json` file)
 - Delete and rerun Canada 2011 geographies since the name of the columns will change (to fix the geonames, which were broken).

javitonino commented on tasks/ca/statcan/data.py:
```diff
@@ -431,5 +441,5 @@ class NHSMetaWrapper(MetaWrapper):
 
     def tables(self):
-        yield Geography(resolution=self.resolution)
+        yield Geography(resolution=self.resolution, year='2011')
```
In general, why are years passed as strings instead of integers?

javitonino commented on tasks/ca/statcan/geo.py:
```diff
@@ -99,20 +89,25 @@
+    # http://www12.statcan.gc.ca/census-recensement/2011/geo/bound-limit/bound-limit-2011-eng.cfm
+    '2011': 'http://www12.statcan.gc.ca/census-recensement/{year}/geo/bound-limit/files-fichiers/g{resolution}000b11{format}_e.zip',
+    # https://www12.statcan.gc.ca/census-recensement/2011/geo/bound-limit/bound-limit-2016-eng.cfm
+    '2016': 'http://www12.statcan.gc.ca/census-recensement/2011/geo/bound-limit/files-fichiers/{year}/l{resolution}000b16a_e.zip',
```
Haha, the URL is for 2011 but then for 2016. xD

javitonino commented on tasks/ca/statcan/geo.py:
```diff
@@ -99,20 +89,25 @@
-                               resolution=self.resolution,
-                               format=GEOGRAPHY_FORMAT[self.resolution])
+        url = YEAR_URL[self.year]
+        return url.format(year=self.year,
```
Maybe we can remove the `year` placeholder since the URL is different per-year anyway.

javitonino commented on tasks/ca/statcan/geo.py:
```diff
@@ -137,15 +133,17 @@ def input_files(self):
     def requires(self):
-        extension = 'gml' if self.resolution == GEO_FSA else 'shp'
-        return ImportGeography(resolution=self.resolution, extension=extension)
+        extension = 'gml' if self.resolution == GEO_FSA and self.year == '2011' else 'shp'
```
There is a `GEOGRAPHY_FORMAT` constant that seems to be somewhat related to this. Could they be unified?

antoniocarlon commented on tasks/ca/statcan/census2016/census_columns_filter.json:
```diff
@@ -0,0 +1,34 @@
+{
```
Nope, just a good text editor and some editing tricks to avoid dying in the process. I thought about creating some tooling but as it would be only valid for Canada 2016 (only one shot), I preferred not to do it

antoniocarlon commented on tasks/ca/statcan/census2016/data.py:
```diff
@@ -0,0 +1,251 @@
+                            measure_names += [colname + '_m']
+                            measure_values += [safe_float_cast(line[12])]
+
+                    measurements = dict(zip(measure_names, measure_values))
```
True. This comes from the fact that in previous versions I was generating all the names first and then the measurements and I changed it when I hit the Postgres column limit and I had to split `languages` in three different tables. I'll give it a try :+1:

antoniocarlon commented on tasks/ca/statcan/census2016/data.py:
```diff
@@ -0,0 +1,251 @@
+                    f_column_name = line[8]
+                    f_measurement_total = safe_float_cast(line[11])
+
+                    LOGGER.debug('Reading line {} ::: {} ({}) | {} | {} '.format(
```
I'll take a look at it

antoniocarlon commented:
Blocked, adding Dissemination Blocks https://github.com/CartoDB/Geographica-Product-Coordination/issues/27#issuecomment-423216287

antoniocarlon commented:
Added Dissemination Blocks interpolation from Dissemination Areas and fixed a small issue with the column naming among source data files --> Back to CR!

juanignaciosl commented:
`docker-ca-all-2016` has been running for more than a day now both in local and staging, it's too slow. Let's improve this.

Here's the duration of the census data tasks:

```
9:35:23 done ca.statcan.census2016.data.ImportData(geocode=044, topic=t004, segment=all)
0:30:26 done ca.statcan.census2016.data.CensusData(resolution=da_, topic=t004, segment=all)
0:26:59 done ca.statcan.census2016.data.CensusDBFromDA(topic=t004, segment=all)
0:00:00 done ca.statcan.census2016.data.CensusDataWrapper(resolution=db_, topic=t004)
3:05:42 done ca.statcan.census2016.data.ImportData(geocode=044, topic=t001, segment=all)
0:08:07 done ca.statcan.census2016.data.CensusData(resolution=da_, topic=t001, segment=all)
0:20:42 done ca.statcan.census2016.data.CensusDBFromDA(topic=t001, segment=all)
0:00:00 done ca.statcan.census2016.data.CensusDataWrapper(resolution=db_, topic=t001)
4:33:22 done ca.statcan.census2016.data.ImportData(geocode=044, topic=t005, segment=all)
0:11:41 done ca.statcan.census2016.data.CensusData(resolution=da_, topic=t005, segment=all)
0:24:53 done ca.statcan.census2016.data.CensusDBFromDA(topic=t005, segment=all)
0:00:00 done ca.statcan.census2016.data.CensusDataWrapper(resolution=db_, topic=t005)
8:16:57 done ca.statcan.census2016.data.ImportData(geocode=044, topic=t006, segment=all)
0:29:41 done ca.statcan.census2016.data.CensusData(resolution=da_, topic=t006, segment=all)
0:24:42 done ca.statcan.census2016.data.CensusDBFromDA(topic=t006, segment=all)
0:00:00 done ca.statcan.census2016.data.CensusDataWrapper(resolution=db_, topic=t006)
4:43:10 done ca.statcan.census2016.data.ImportData(geocode=044, topic=t002, segment=all)
0:15:09 done ca.statcan.census2016.data.CensusData(resolution=da_, topic=t002, segment=all)
0:21:56 done ca.statcan.census2016.data.CensusDBFromDA(topic=t002, segment=all)
0:00:00 done ca.statcan.census2016.data.CensusDataWrapper(resolution=db_, topic=t002)
```

`ImportData` is just too slow. It spends ages on queries like this one:

```
postgres10_1          | 	                           INSERT INTO "ca.statcan.census2016.data".ImportData_044_m_t003_f9831b9612 (geom_id, geolevel, c0335_m)
postgres10_1          | 	                           SELECT '2442020', 3, NULLIF('0.0', 'x')::NUMERIC
postgres10_1          | 	                           ON CONFLICT (geom_id)
postgres10_1          | 	                           DO UPDATE SET c0335_m = EXCLUDED.c0335_m
```

antoniocarlon commented:
Blocked by https://github.com/CartoDB/bigmetadata/pull/584
